test.rb

```diff
@@ -0,0 +1,251 @@
+            total_maths += element[:maths]
+            total_chemistry += element[:chemistry]
+            print "|   #{element[element.keys[0]]}   |"
+            display_fields.each do |field|
+                print element[field.to_sym].to_s + '|'
+            end
+            print "\n"
+        end # loop end
+
+        if show_total.to_s.casecmp('TRUE').zero?
+            print ' |   Total   '
+
+            display_fields.each do |field|
+                if field == 'maths'
+                    print " |    #{total_maths}       "
+                elsif field == 'physics'
+                    print " |    #{total_physics}       "
+                else
+                    print "|    #{total_chemistry}      "
+                end
+            end
+            print "\n"
+       end
+    end # display_student_data ends
+
+    #  Should compare Hashes method starts
+    def student_compare_data(compare_on, first_comprer_element, second_comparer_element, group_by, show_total)
+        @hash_array = @hash_array.sort_by { |hsh| hsh[:year] }.reverse
+        temp_arr = []
+        temp_year = @hash_array[0][:year]
+        final_year = 0
+        (0...@hash_array.size).each do |iter|
+            if  @hash_array[iter][:year] == temp_year
+                temp_arr.push(@hash_array[iter][:student_id])
+            else
+                temp_arr.delete(@hash_array[iter][:student_id])
+                final_year = @hash_array[iter][:year]
+            end
+        end
+        array_push(student_id: temp_arr[0], department: 'a7', maths: 0, physics: 0, chemistry: 0, year: final_year)
+
+        # Main if starts
+        if compare_on.casecmp('year').zero?
+            @hash_array = @hash_array.sort_by { |hsh| hsh[:student_id] }
+            display_compare_student_data(first_comprer_element, second_comparer_element, group_by, show_total)
+        end
+        # Main if  ends
+    end
+    #  Should compare Hashes method ends
+
+    #  display_compare_student_data method starts
+    def display_compare_student_data(first_comprer_element, second_comparer_element, group_by, show_total)
+        puts " |   #{group_by}   |    MATHS    |    PHYSICS    |    CHEMISTRY    |"
+        temp = 0
+        math_curr = 0
+        physics_curr = 0
+        chemistry_curr = 0
+        math_prev = 0
+        physics_prev = 0
+        chemistry_prev = 0
+        math_first_comprer_total = 0
+        math_second_comprer_total = 0
+        physics_first_comprer_total = 0
+        physics_second_comprer_total = 0
+        chemistry_first_comprer_total = 0
+        chemistry_second_comprer_total = 0
+        (0...@hash_array.size).each do |iter| # @hash_array.each do |element|
+            if temp == 0 || @hash_array[iter][:student_id] != temp
+                puts " |   #{@hash_array[iter][:student_id]} |"
+                temp = @hash_array[iter][:student_id]
+
+            end # if -else ends
+
+            if @hash_array[iter][:year] == first_comprer_element.to_i
```

Comment:
year = @hash_array[iter][:year] == first_comprer_element.to_i ? first_comprer_element : second_comparer_element

Using year to replace first_comprer_element and second_comparer_element will reduce your code and follows the principle of DRY